client/src/app/shared/components/sorted-tree/sorted-tree.component.ts

```diff
@@ -0,0 +1,630 @@
+    /**
+     * The data to build the tree
+     */
+    public osTreeData: OSFlatNode[] = [];
+
+    /**
+     * The tree control
+     */
+    public treeControl = new FlatTreeControl<OSFlatNode>(node => node.level, node => node.expandable);
+
+    /**
+     * Source for the tree
+     */
+    public dataSource = new ArrayDataSource(this.osTreeData);
+
+    /**
+     * Number to calculate the next position the node is moved
+     */
+    private nextPosition = 0;
+
+    /**
+     * Number which defines the level of the placeholder.
+     * Necessary to show the placeholder for the moved node correctly.
+     */
+    public placeholderLevel = 0;
+
+    /**
+     * Node with calculated next information.
+     * Containing information like the position, when the drag starts and where it is in the moment.
+     */
+    public nextNode: ExFlatNode = null;
+
+    /**
+     * Pointer for the move event
+     */
+    private pointer: DragEvent = null;
+
+    /**
+     * Subscription for the data store
+     */
+    private modelSubscription: Subscription = null;
+
+    /**
+     * Input that defines the key for the parent id
+     */
+    @Input()
+    public parentKey: keyof T;
+
+    /**
+     * Input that defines the key for the weight of the items.
+     * The weight defines the order of the items.
+     */
+    @Input()
+    public weightKey: keyof T;
+
+    /**
+     * Setter to get all models from data store.
+     * It will create or replace the existing subscription.
+     */
+    @Input()
+    public set model(model: Observable<T[]>) {
+        if (!model) {
+            return;
+        }
+
+        if (this.modelSubscription) {
+            this.modelSubscription.unsubscribe();
+        }
+
+        this.modelSubscription = model.pipe(auditTime(10)).subscribe(values => {
+            this.osTreeData = this.treeService.buildFlatTree(values, this.weightKey, this.parentKey);
+            this.dataSource = new ArrayDataSource(this.osTreeData);
+            this.treeControl.expandAll();
+        });
+    }
+
+    /**
+     * Constructor
+     *
+     * @param treeService Service to get data from store and build the tree nodes.
+     */
+    public constructor(private treeService: TreeService) {}
+
+    /**
+     * On init method
+     */
+    public ngOnInit(): void {}
+
+    /**
+     * On destroy - unsubscribe the subscription
+     */
+    public ngOnDestroy(): void {
+        this.modelSubscription.unsubscribe();
+    }
+
+    /**
+     * Function to check if the node has a parent.
+     *
+     * @param node which is viewed.
+     *
+     * @returns The parent node if available otherwise it returns null.
+     */
+    public getParentNode(node: OSFlatNode): OSFlatNode {
+        const nodeIndex = this.osTreeData.indexOf(node);
+
+        for (let i = nodeIndex - 1; i >= 0; --i) {
+            if (this.osTreeData[i].level === node.level - 1) {
```

Comment:
This sometimes throws this error: `SortedTreeComponent.html:10 ERROR TypeError: Cannot read property 'level' of undefined`